azure/functions/meta.py

```diff
@@ -181,37 +181,79 @@ def _parse_timedelta_metadata(
     @classmethod
     def _parse_datetime(
             cls, datetime_str: str) -> datetime.datetime:
+        too_fractional = re.match(
+            r'(.*\.\d{6})(\d+)(Z|[\+|-]\d{1,2}:\d{1,2}){0,1}', datetime_str)
+
+        if too_fractional:
+            # The supplied value contains seven digits in the
+            # fractional second part, whereas Python expects
+            # a maxium of six, so strip it.
+            # https://github.com/Azure/azure-functions-python-worker/issues/269
+            datetime_str = too_fractional.group(1) + (
+                too_fractional.group(3) or '')
+
+        # Try parse time
+        utc_time, utc_time_error = cls._parse_datetime_utc(datetime_str)
+        if utc_time:
+            return utc_time.replace(tzinfo=datetime.timezone.utc)
+
+        local_time, local_time_error = cls._parse_datetime_local(datetime_str)
+        if local_time:
+            return local_time.replace(tzinfo=None)
+
+        # Report error
+        if utc_time_error:
+            raise utc_time_error
+        elif local_time_error:
+            raise local_time_error
+
+    @classmethod
+    def _parse_datetime_utc(
+        cls, datetime_str: str
+    ) -> typing.Tuple[datetime.datetime, Exception]:
+
         # UTC ISO 8601 assumed
-        formats = [
+        # 2018-08-07T23:17:57.461050Z
+        utc_formats = [
             '%Y-%m-%dT%H:%M:%S+00:00',
             '%Y-%m-%dT%H:%M:%S-00:00',
             '%Y-%m-%dT%H:%M:%S.%f+00:00',
             '%Y-%m-%dT%H:%M:%S.%f-00:00',
             '%Y-%m-%dT%H:%M:%SZ',
             '%Y-%m-%dT%H:%M:%S.%fZ',
         ]
+
         dt = None
+        last_error = None
+        for utc_fmt in utc_formats:
+            try:
+                dt = datetime.datetime.strptime(datetime_str, utc_fmt)
+            except ValueError as e:
+                last_error = e
 
-        too_fractional = re.match(
-            r'(.*\.\d{6})(\d+)(Z|[\+|-]\d{1,2}:\d{1,2})', datetime_str)
+        return dt, last_error
 
-        if too_fractional:
-            # The supplied value contains seven digits in the
-            # fractional second part, whereas Python expects
-            # a maxium of six, so strip it.
-            # https://github.com/Azure/azure-functions-python-worker/issues/269
-            datetime_str = too_fractional.group(1) + too_fractional.group(3)
+    @classmethod
+    def _parse_datetime_local(
+        cls, datetime_str: str
+    ) -> typing.Tuple[datetime.datetime, Exception]:
+
+        # Local time assumed
+        # 2018-08-07T23:17:57.461050
+        local_formats = [
+            '%Y-%m-%dT%H:%M:%S.%f',
+            '%Y-%m-%dT%H:%M:%S',
+        ]
 
-        for fmt in formats:
+        dt = None
```

Comment:
nit: I'd refactor the `_parse_datetime_local` and `_parse_datetime_utc` functions to use a third function that encapsulates the `striptime` logic since that is the same across both. The only variation is the formats which could be passed in as separated arguments.